Rewrite and fix change_history logic

Chrislee187 opened the pull request:
https://trello.com/c/oTNmyzgu/146-rewrite-and-fix-change-history-logic-in-spv2-medium

Commit:
No longer added first published note to drafts

Commit:
Added First published change-history note when first publishing a document

Commit:
Localised change_history code

* Set timezone of app to London as that's where the app
  will be run, and doing so will help avoid serialization issues
  with timestamps ('Z' vs. '+00:00').
* Moved change_history from DocumentPresenter to Document (model)
* removed STUB_TIME_STAMP now uses current time
* stringified keys for new "first published" entry in change_history as
  when they come from the API they'll all be string keys so this
  makes API vs. new ones more consistent.

tuzz commented on app/models/document.rb:
```diff
@@ -208,6 +214,12 @@ def save
+      if not_published?
+        self.change_note = "First published."
+        self.save
+      end
```
I think we need to be careful with the `#not_published?` method. It returns true if the `publication_state` returned by Publishing API is not one of `live` or `redrafted`. The Publishing API can also return states of `republish` and `unpublish` I think.

I'm not sure what the right thing to do is, though. In the case of `republish` it's ambiguous as to whether the content item was previously published or not. It might be safer to look through the `change_history` for the "First published." item or maybe look at the `first_published_at` field which should always be set on a content item that has previously been published.

h-lame replied:
Some document formats (dfid research output) allow editing `first_published_at` directly so checking for that doesn't seem safe.  Checking the current state for any value doesn't seem safe either because the transitions aren't linear (e.g. we can go from draft to unpublished without going through published).  The "safest" option seems to be checking for "First published" in the change note (the edgecase here would be some editor adding that as their change note before it actually was published, but that'd be pretty pathological).

Hopefully we can do https://trello.com/c/kyRoUV0h/158-add-a-state-history-field-to-presented-items-in-publishing-api-medium soon and can re-write `ever_been_published?` to interrogate the full state history.  I'll add a note to https://trello.com/c/LU6PckOO/182-fix-document-history-for-redrafts-of-unpublished-documents to make it clear that this method would need to be changed.

How does that sound?

h-lame replied:
Oh, this mysterious `ever_been_published?` method I mention is one I propose to write in this PR to do the search through change history for "First published", instead of using `not_published?` in this conditional.

tuzz replied:
I think I'd independently arrived at the same conclusion, this approach seems most accurate and as you say, once we have the 'state_history' information we can make this cleaner.

tuzz commented on app/presenters/document_presenter.rb:
```diff
@@ -75,13 +75,16 @@ def metadata
+  #   when "minor"
+  #     document.change_history
+  #   end
+  # end
```
delete

tuzz commented on config/application.rb:
```diff
@@ -24,7 +22,7 @@ class Application < Rails::Application
     # Run "rake -D time" for a list of tasks for finding time zone names. Default is UTC.
     # config.time_zone = 'Central Time (US & Canada)'
-
+    config.time_zone = 'London'
```
👍

h-lame committed:
Ensure we have predicates for all publishing-api states

The possible states are:
  draft
  live
  redrafted
  superseded
  unpublished

This is from looking at all states in the publishing-api db, and
comparing with the presenter code[1] to work out what possible
states could be returned.  We didn't have superseded? or unpublished?
so I've added them.

[1]: https://github.com/alphagov/publishing-api/blob/master/app/presenters/queries/content_item_presenter.rb#L114-L124

h-lame committed:
Always extract change_history in from_publishing_api

Prior to this change we would only extract the change_history if the
document was what we considered "published" (e.g. "live" or "rerafted").
If we don't extract the change_history from the API payload then when
we write the document back to the publishing-api we will clear out the
change_history.  This feels overly destructive and something that we
should take a specific and explicit step to achieve if we do want it
rather than the current approach which feels like it might be an
oversight.

h-lame committed:
Add Document#has_ever_been_published?

The current implementation is to look for the presence of the "First
published." note that we add during publish! - if we find it then we
assume the document has been published before, if we don't then we
assume it hasn't.  We then use this new method to decide if we should add
the "First published." note during the call to publish!.

This isn't the best way to decide if a document has ever been published
but the alternatives all have problems.  first_published_at is one
candidate as it is added by publishing-api on first publication, but some
formats (e.g. dfid research output) allow editing this field before
publishing so we can't rely on it.  We also can't use the state of the
document because that can only tell us if the document is currently
published not if it has ever been published (for example a document can
go straight from draft to unpublished without going through published
beforehand).

There is a story to play soon that will add full state_history to the
response from the publishing-api at which point we'll update the
implementation of has_ever_been_published? to use this information and
be more robust.

h-lame committed:
Add published trait for document factories

Our understanding of what a "published" document is has changed a bit
now that we consider the contents of its change_history as marker. In
order that our tests are more closely aligned with real data we introduce
a published trait for the document factory that makes sure the expected
entires are in the change_history.  We also make change_history an
attribute that can be specified without digging deep into the details
hash.

We've updated some of the specs to use this new trait.
